agents/magpie/magpie_agent.py

```diff
@@ -0,0 +1,388 @@
+            fp.ys[i] = y * wafer_scale
+            fp.rots[i] = np.deg2rad(pol_angle)
+            fp.cnames[i] = f"chan_{abschan}"
+            fp.eqs[i] = eq.format(fp.cnames[i])
+            fp.cmaps[i] = cmaps[freq]
+            fp.templates[i] = templates[freq]
+        map_inv = np.array(map_inv)
+        # Mapping form channel no. to fp-idx
+        fp.chan_mask = np.full(np.max(map_inv)+1, -1)
+        for idx, chan in enumerate(map_inv):
+            fp.chan_mask[chan] = idx
+
+        return fp
+
+    @classmethod
+    def from_wafer_file(cls, wafer_file, wafer_scale=50):
+        import copy
+        import quaternionarray as qa
+        import toml
+
+        dets = toml.load(wafer_file)
+        fp = cls(ndets=len(dets))
+        band_idxs = {}
+        bands_seen = 0
+
+        xaxis = np.array([1., 0., 0.])
+        zaxis = np.array([0., 0., 1.])
+        cmaps = ["red_cmap", "blue_cmap"]
+
+        def det_coords(det):
+            quat =np.array(det['quat']).astype(np.float64)
+            rdir = qa.rotate(quat, zaxis).flatten()
+            ang = np.arctan2(rdir[1], rdir[0])
+            orient = qa.rotate(quat, xaxis).flatten()
+            polang = np.arctan2(orient[1], orient[0])
+            mag = np.arccos(rdir[2]) * 180 / np.pi
+            xpos = mag * np.cos(ang)
+            ypos = mag * np.sin(ang)
+            return (xpos, ypos), polang
+
+        for key, det in dets.items():
+            chan = det['channel']
+            (x, y), polangle = det_coords(det)
+
+            if det['band'] not in band_idxs:
+                band_idxs[det['band']] = bands_seen
+                bands_seen += 1
+            color_idx = band_idxs[det['band']]
+
+            fp.xs[chan] = x * wafer_scale
+            fp.ys[chan] = y * wafer_scale
+            fp.rots[chan] = polangle
+            fp.cnames[chan] = key
+            fp.eqs[chan] = f"/ + 1 s {key} 2"
+            fp.cmaps[chan] = cmaps[color_idx]
+            fp.templates[chan] = f"template_c{color_idx}_p0"
+
+        return fp
+
+
+
+class MagpieAgent:
+    mask_register = 'AMCc.SmurfProcessor.ChannelMapper.Mask'
+
+    def __init__(self, agent, args):
+        self.agent = agent
+        self._running = False
+
+        self.target_rate = args.target_rate
+        layout = args.layout.lower()
+        if layout == 'grid':
+            self.fp = FocalplaneConfig.grid(
+                args.xdim, args.ydim, ygap=8
+            )
+        elif layout == 'wafer':
+            if args.csv_file is not None:
+                self.fp = FocalplaneConfig.from_tune_csv(
+                    args.csv_file, wafer_scale=args.wafer_scale
+                )
+            else:
+                self.fp = FocalplaneConfig.from_wafer_file(
+                    args.wafer_file, wafer_scale=args.wafer_scale
+                )
+
+        self.sent_cfg = False
+        self.num_dets = self.fp.num_dets
+        self.mask = np.arange(self.num_dets)
+        self.mask = np.arange(4960)
```

Comment:
This looks like it's supposed to be 4096 ... and like it trashes the line before so not both are needed.